rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+    init_with_barrier(mem, &mut (*region).vec_pages, vec_pages);
+
+    r_ptr
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_id<M: Memory>(_mem: &mut M, r: Value) -> u32 {
+    let r = r.as_region();
+    (*r).id.into()
+}
+
+// Region manager's total memory size in stable memory, in _pages_.
+#[ic_mem_fn]
+pub unsafe fn region_get_mem_size<M: Memory>(_mem: &mut M) -> u64 {
+    let size = {
+        if let Some(s) = crate::region::REGION_SET_MEM_SIZE {
+            return s;
+        };
+        if crate::region::REGION_MEM_SIZE_INIT {
+            meta_data::size::STATIC_MEM_IN_PAGES as u64
+                + crate::region::REGION_TOTAL_ALLOCATED_BLOCKS as u64
+                    * (meta_data::size::PAGES_IN_BLOCK as u64)
+        } else {
+            // Before initialization of anything, give back zero.
+            0
+        }
+    };
+    size
+}
+
+// Region manager's total memory size in stable memory, in _pages_.
+#[ic_mem_fn]
+pub unsafe fn region_set_mem_size<M: Memory>(_mem: &mut M, size: u64) {
+    crate::region::REGION_SET_MEM_SIZE = Some(size);
+}
+
+// Helper for commmon logic that reserves low-valued RegionIds in a certain span for future use.
+// When first is some, we are actually reserving.  When first is none, we are checking that the reservation has occured.
+unsafe fn region_reserve_id_span<M: Memory>(_mem: &mut M, first: Option<RegionId>, last: RegionId) {
+    if let Some(first) = first {
+        let next_id = meta_data::total_allocated_regions::get() as u16;
+        assert_eq!(first.0, next_id);
+        assert!(first.0 <= last.0);
+        meta_data::total_allocated_regions::set((last.0 + 1) as u64);
+    } else {
+        let next_id = meta_data::total_allocated_regions::get() as u16;
+        assert!(last.0 < next_id);
+    }
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_new<M: Memory>(mem: &mut M) -> Value {
+    let next_id = meta_data::total_allocated_regions::get() as u16;
+    meta_data::total_allocated_regions::set(next_id as u64 + 1);
+
+    let vec_pages = alloc_blob(mem, Bytes(0));
+    let r_ptr = alloc_region(mem, next_id, 0, vec_pages);
+
+    // Update Region table.
+    {
+        let r_id = RegionId::from_id(next_id);
+        let c = meta_data::region_table::get(&r_id);
+        // sanity check: Region table says that this region is available.
+        assert_eq!(c, None);
+        meta_data::region_table::set(&r_id, Some(RegionSizeInPages(0)));
+    }
+    r_ptr
+}
+
+pub unsafe fn region_recover<M: Memory>(mem: &mut M, rid: &RegionId) -> Value {
+    use meta_data::size::PAGES_IN_BLOCK;
+    let c = meta_data::region_table::get(&rid);
+    let page_count = match c {
+        Some(RegionSizeInPages(pc)) => pc,
+        _ => {
+            rts_trap_with("region_recover failed: Expected Some(RegionSizeInPages(_))");
+        }
+    };
+
+    let block_count = (page_count as u32 + PAGES_IN_BLOCK - 1) / PAGES_IN_BLOCK;
+    let vec_pages = alloc_blob(mem, Bytes(block_count * 2));
```

Comment:
I am curious about the `* 2`. Is it to reserve some capacity for further region growth?

Reply:
I think it's because each block id in the vec is a u16, so 2 bytes.

Reply:
True, thanks. Maybe, a `size_of::<u16>()` or something like `size_of::<BlockId>()` (and introducing a `type BlockId = u16`) could be helpful to make it more self-explanatory.